### ISSUE TITLE
Win IoT Core/RPi3 GetTwinAsync infinite retry/failure loop

### ISSUE THREAD

**Issue by zivkan** [Azure/azure-iot-sdk-csharp#1460]
# Context

- **OS, version, SKU and CPU architecture used:** Windows 10 IoT Core arm64
- **Application's .NET Target Framework :** netcoreapp3.1.
- **Device:** Raspberry PI3
- **SDK version used:** 3.1.301,
- **PackageReferences**
  - `<PackageReference Include="Microsoft.Azure.Devices.Client" Version="1.27.0" />`

## Description of the issue
On the Raspberry Pi, running Windows IoT Core, `GetTwinAsync` never returns, and the connection status changes handler keeps getting events saying that it's connected then there was a communications error. I left it running for a few minutes to see if it will eventually work, but it did not, and when I checked my IoT Hub, it had used 1000 of the 8000 messages on the free pricing tier.

The exact same code works on my devbox without any reconnections, it's only an issue on the RPi. Both devices are on the same network, both wired.

## Code sample exhibiting the issue

```cs
using System;
using System.Threading.Tasks;
using Microsoft.Azure.Devices.Client;
using Microsoft.Azure.Devices.Shared;

class Program
{
    static async Task Main(string[] args)
    {
        var hostname = args[0];
        var deviceId = args[1];
        var sasToken = args[2];

        var client = DeviceClient.Create(hostname,
            AuthenticationMethodFactory.CreateAuthenticationWithToken(deviceId, sasToken),
            TransportType.Mqtt);

        client.SetConnectionStatusChangesHandler(ConnectionStatusChangesHandler);

        Twin twin = await client.GetTwinAsync();
    }

    private static void ConnectionStatusChangesHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
    {
        var msg = DateTime.UtcNow.ToString("O") + " " + status + " " + reason;
        Console.WriteLine(msg);
    }
}
```

## Console log of the issue
Setting the DEBUG environment variable to '*' didn't output anything other than my own app's `Console.WriteLine` output.

```text
2020-07-11T15:02:02.5023394Z Connected Connection_Ok
2020-07-11T15:02:03.1838421Z Disconnected_Retrying Communication_Error
2020-07-11T15:02:04.2866074Z Connected Connection_Ok
2020-07-11T15:02:04.6673275Z Disconnected_Retrying Communication_Error
2020-07-11T15:02:05.7413474Z Connected Connection_Ok
2020-07-11T15:02:06.2766288Z Disconnected_Retrying Communication_Error
2020-07-11T15:02:07.3932190Z Connected Connection_Ok
2020-07-11T15:02:07.8707301Z Disconnected_Retrying Communication_Error
^C
```

**Comment by zivkan** [comment]
Other things I should have mentioned:

The same problem exists with the UWP app that I'm trying to develop to display on the official RPi touchscreen.

I enabled the Azure Device Management client on the Raspberry Pi, and now when I power cycle the device, IoT Hub registers 15 more messages. Looking at the \\device\c$\dm directory, I can see the logs are all successful, so it doesn't appear that then device management client has the same problems getting or the device twin's desired state.

**Comment by abhipsaMisra** [comment]
Could you capture the SDK logs and share them with us? You should be able to use `tracelog`, or any other event tracing tool, to do that.
`tracelog -start IotTrace -f iot.etl -guid iot_providers.txt`

You can find the SDK event providers here: https://github.com/Azure/azure-iot-sdk-csharp/tree/master/tools/CaptureLogs#azure-iot-sdk-providers

**Comment by zivkan** [comment]
I used the console logger, and took out my IoT Hub hostname and device name. I don't think there's anything else PII in the log
[log.txt](https://github.com/Azure/azure-iot-sdk-csharp/files/4942097/out.txt)

It loos like the error is

> 2020-07-18T06:31:45.0833414 [Microsoft-Azure-Devices-Device-Client-Enter] (MqttIotHubAdapter#31884011, ShutdownOnError, (MqttIotHubAdapter#0, System.Net.Sockets.SocketException (10054): An existing connection was forcibly closed by the remote host.)).

As someone who also working a .NET app that does networking, I can understand that this is difficult (impossible?) to diagnose. Our apps just use the .NET APIs for networking, and it's up to the implementation of .NET and the OS's TCP stack and the network driver, and the internet as a whole to work properly.  But only my RPi has this problem, my desktop/laptops on the same network do not have this problem.   Are you aware of any networking problems with Windows IoT Core on Raspberry Pi 3?  Any tips on what else I can investigate?

Although the Device Management app doesn't seem to report any problems, and my UWP app posts to an Azure Functions app every 60 seconds, and that works just fine.

**Comment by bikamani** [comment]
@zivkan Thanks for your patience. 
Yes logs indicate network exception. I am curious to know if you are still seeing same behavior on latest version of our device SDK? Have you tried other protocols?

**Comment by bikamani** [comment]
> @zivkan Thanks for your patience.
> Yes logs indicate network exception. I am curious to know if you are still seeing same behavior on latest version of our device SDK? Have you tried other protocols?

To answer your previous question - I am not aware of networking issues on Raspberry Pi 3.

**Comment by zivkan** [comment]
I just tried again with package version 1.33.0, and it seems to be working now, at least from a console app. I haven't worked on my UWP app since July, but I'll try it at some point. Thanks.

**Comment by az-iot-builder-01** [comment]
@zivkan, @abhipsaMisra, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by bikamani** [comment]
Thanks for letting us know @zivkan .

**Comment by v-tbert** [comment]
FWIW I experienced a similar problem yesterday with 1.33.1. I could connect to IoTHub just fine but if I tried to do anything, like set method handlers or send property updates it would go into a rapid, endless reconnect cycle with status Disconnected_Retrying Communication_Error. What I eventually realized was that I had the process running on another test machine with the same iothub credentials. Apparently, using duplicate credentials lets you connect but using the connection causes iothub backend to drop the connection with an exception. That exception gets swallowed internally somewhere in device client code obfuscating the real cause.
